c++/src/capnp/cc_capnp_library.bzl

```diff
@@ -0,0 +1,119 @@
+
+    out_dir = ctx.var["GENDIR"]
+    if src_prefix != "":
+        out_dir = out_dir + "/" + src_prefix
+
+    cc_out = "-o%s:%s" % (ctx.executable._capnpc_cxx.path, out_dir)
+    args = ctx.actions.args()
+    args.add_all(["compile", "--verbose", cc_out])
+    args.add_all(["-I" + inc for inc in includes])
+    args.add_all(["-I", system_include])
+    if src_prefix != "":
+        args.add_all(["--src-prefix", src_prefix])
+
+    args.add_all([s for s in ctx.files.srcs])
+
+    ctx.actions.run(
+        inputs = inputs + ctx.files._capnpc_cxx + ctx.files._capnpc_capnp + ctx.files._capnp_system,
+        outputs = ctx.outputs.outs,
+        executable = ctx.executable._capnpc,
+        arguments = [args],
+        mnemonic = "GenCapnp",
+    )
+
+    return [
+        capnp_provider(
+            includes = includes,
+            inputs = inputs,
+            src_prefix = src_prefix,
+        ),
+    ]
+
+_capnp_gen = rule(
+    attrs = {
+        "srcs": attr.label_list(allow_files = True),
+        "deps": attr.label_list(providers = [capnp_provider]),
+        "data": attr.label_list(allow_files = True),
+        "outs": attr.output_list(),
+        "src_prefix": attr.string(),
+        "_capnpc": attr.label(executable = True, allow_single_file = True, cfg = "exec", default = "@capnp-cpp//src/capnp:capnp_tool"),
+        "_capnpc_cxx": attr.label(executable = True, allow_single_file = True, cfg = "exec", default = "@capnp-cpp//src/capnp:capnpc-c++"),
+        "_capnpc_capnp": attr.label(executable = True, allow_single_file = True, cfg = "exec", default = "@capnp-cpp//src/capnp:capnpc-capnp"),
+        "_capnp_system": attr.label(default = "@capnp-cpp//src/capnp:capnp_system_library"),
+    },
+    output_to_genfiles = True,
+    implementation = _impl,
+)
+
+def cc_capnp_library(
+        name,
+        srcs = [],
+        data = [],
+        deps = [],
+        src_prefix = "",
+        visibility = None,
+        target_compatible_with = None,
+        **kargs):
+    """Bazel rule to create a C++ capnproto library from capnp source files
+
+    Args:
+        name: library name
+        srcs: list of files to compile
+        data: additional files to provide to the compiler - data files and includes that need not to
+            be compiled
+        deps: other cc_capnp_library rules to depend on
+        src_prefix: src_prefix for capnp compiler to the source root
+        visibility: rule visibility
+        target_compatible_with: target compatibility
+        **kargs: rest of the arguments to cc_library rule
+    """
+
+    hdrs = [s + ".h" for s in srcs]
+    srcs_cpp = [s + ".c++" for s in srcs]
+
+    _capnp_gen(
+        name = name + "_gen",
+        srcs = srcs,
+        deps = [s + "_gen" for s in deps],
+        data = data,
+        outs = hdrs + srcs_cpp,
+        src_prefix = src_prefix,
+        visibility = visibility,
+        target_compatible_with = target_compatible_with,
+    )
+    native.cc_library(
```

Comment:
Why not just use [CcInfo](https://docs.bazel.build/versions/main/integrating-with-rules-cc.html#implementing-starlark-rules-that-depend-on-c-rules-andor-that-c-rules-can-depend-on)?

example from rules_go: https://github.com/bazelbuild/rules_go/blob/df02d01fb37701ff826697271220f043b364d1c1/go/private/rules/binary.bzl#L72

Reply:
I'm not sure what that is, it will take me some time to figure it out. Can I postpone this to a future PR?

Also what are the benefits of it? It is a cc_library anyway.